### PR TITLE
Add SMS OTP as a service

### PR DESCRIPTION
**Pull request by nipunthathsara**
Git issue : [https://github.com/wso2/product-is/issues/12328](https://github.com/wso2/product-is/issues/12328)

## Purpose
To expose the SMS OTP generation and validation capability as a fully decoupled OSGI service.

### CODE REVIEW

File: component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/util/Utils.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ */
+public class Utils {
+
+    public static Properties readConfigurations() throws SMSOTPServerException {
+
+        try {
+            ModuleConfiguration configs = IdentityEventConfigBuilder.getInstance()
+                    .getModuleConfigurations(Constants.SMS_OTP_IDENTITY_EVENT_MODULE_NAME);
+            // Work with the default values if configurations couldn't be loaded.
+            return configs != null ? configs.getModuleProperties() : new Properties();
```
Comment (somindatommy) [inline component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/util/Utils.java:40]: Don't we need to add an error log saying configurations cannot be loaded?

File: component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/SMSOTPServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+        if (StringUtils.isBlank(userId)) {
+            throw Utils.handleClientException(Constants.ErrorMessage.CLIENT_EMPTY_USER_ID, null);
+        }
+        // Retrieve user by ID.
+        AbstractUserStoreManager userStoreManager;
+        User user;
+        try {
+            userStoreManager = (AbstractUserStoreManager) SMSOTPServiceDataHolder.getInstance()
+                    .getRealmService().getTenantUserRealm(getTenantId()).getUserStoreManager();
+            user = userStoreManager.getUser(userId, null);
```
[the remark this answers is not in the document]
Comment (nipunthathsara) [reply component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/SMSOTPServiceImpl.java:73]: `userStoreManager` won't be null, and there's a null check for the `user`.

File: component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/SMSOTPServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+                log.debug(String.format("Provided transaction Id doesn't match. User : %s.", userId));
+            }
+            error = showFailureReason ?
+                    new ErrorDTO(Constants.ErrorMessage.CLIENT_INVALID_TRANSACTION_ID, transactionId) : null;
+            return new ValidationResponseDTO(userId, false, error);
+        }
+        return new ValidationResponseDTO(userId, true);
+    }
+
+    private SessionDTO proceedWithOTP(String mobileNumber, User user) throws SMSOTPException {
```
[the remark this answers is not in the document]
Comment (nipunthathsara) [reply component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/SMSOTPServiceImpl.java:201]: `proceedWithOTP()` refactored to `generateOTP `.  `triggerNotification()` is already decoupled.

File: component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/util/Utils.java (hunk 1 above)
Comment (nipunthathsara) [reply component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/util/Utils.java:40]: Bundle activation will not happen as `enabled` property cannot be found.  No exception thrown as per the requirement of having the jar in place but without any configs, feature cannot be used.

File: component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/SMSOTPServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+        // Retrieve user by ID.
+        AbstractUserStoreManager userStoreManager;
+        User user;
+        try {
+            userStoreManager = (AbstractUserStoreManager) SMSOTPServiceDataHolder.getInstance()
+                    .getRealmService().getTenantUserRealm(getTenantId()).getUserStoreManager();
+            user = userStoreManager.getUser(userId, null);
+        } catch (UserStoreException e) {
+            // Handle user not found.
+            if ("30007".equals(((org.wso2.carbon.user.core.UserStoreException) e).getErrorCode())) {
```
Comment (omindu) [inline component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/SMSOTPServiceImpl.java:82]: Use a meaningful const

File: component/common/pom.xml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                            org.apache.commons.logging.*; version="1.0.4",
+                            org.osgi.framework,
+                            org.osgi.service.component; version="${osgi.service.import.version.range}",
+                            *;resolution:=optional
+                        </Import-Package>
+                        <Export-Package>
+                            !org.wso2.carbon.identity.smsotp.common.internal,
+                            org.wso2.carbon.identity.smsotp.common.*; version="${project.version}",
+                        </Export-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (AnuradhaSK) [inline component/common/pom.xml:106]: Remove this Dynamic import, and define all required packages under  Import-Package

File: component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/dto/FailureReasonDTO.java [hunk 6]
```diff
@@ -18,7 +18,7 @@
 
 package org.wso2.carbon.identity.smsotp.common.dto;
 
-import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.StringUtils;
```
Comment (somindatommy) [inline component/common/src/main/java/org/wso2/carbon/identity/smsotp/common/dto/FailureReasonDTO.java:21]: Is this needed?

**Review by vihanga-liyanage — APPROVED** [review]

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**